ledger/ledger-api-integration-tests/src/test/itsuite/scala/com/digitalasset/platform/tests/integration/ledger/api/ContractKeysIT.scala

```diff
@@ -0,0 +1,465 @@
+// Copyright (c) 2019 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.digitalasset.platform.tests.integration.ledger.api
+
+import java.util.UUID
+
+import com.digitalasset.daml.lf.language.{
+  LanguageVersion,
+  LanguageMajorVersion => LMajV,
+  LanguageMinorVersion => LMinV
+}
+import com.digitalasset.ledger.api.testing.utils.{
+  AkkaBeforeAndAfterAll,
+  SuiteResourceManagementAroundEach
+}
+import com.digitalasset.ledger.api.v1.value.{Optional, Record, RecordField, Value}
+import com.digitalasset.platform.apitesting.LedgerContextExtensions._
+import com.digitalasset.platform.apitesting.TestParties._
+import com.digitalasset.platform.apitesting.{MultiLedgerFixture, TestIdsGenerator, TestTemplateIds}
+import com.digitalasset.platform.participant.util.ValueConversions._
+import com.digitalasset.daml.lf.transaction.VersionTimeline.Implicits._
+import com.google.rpc.code.Code
+import org.scalatest.concurrent.{AsyncTimeLimitedTests, ScalaFutures}
+import org.scalatest.{AsyncWordSpec, Matchers, OptionValues}
+
+class ContractKeysIT
+    extends AsyncWordSpec
+    with AkkaBeforeAndAfterAll
+    with MultiLedgerFixture
+    with SuiteResourceManagementAroundEach
+    with ScalaFutures
+    with AsyncTimeLimitedTests
+    with Matchers
+    with OptionValues {
+
+  protected val testTemplateIds = new TestTemplateIds(config)
+  protected val templateIds = testTemplateIds.templateIds
+  protected val testIdsGenerator = new TestIdsGenerator(config)
+
+  private lazy val dummyTemplates =
+    List(templateIds.dummy, templateIds.dummyFactory, templateIds.dummyWithParam)
+  private val operator = Grace
+  private val receiver = Heidi
+  private val giver = Alice
+  private val owner = Bob
+  private val delegate = Charlie
+  private val observers = List(Eve, Frank)
+
+  private val emptyRecordValue = Value(Value.Sum.Record(Record()))
+
+  /** This should be the same version as the dar version. */
+  protected val languageVersion: LanguageVersion = LanguageVersion(LMajV.V1, LMinV.Dev)
+
+  override protected def config: Config = Config.default
+
+  "Working with contract keys" should {
+
+    "permit fetching a divulged contract" in allFixtures { ctx =>
+      def pf(label: String, party: String) =
+        RecordField(label, Some(Value(Value.Sum.Party(party))))
+
+      // TODO currently we run multiple suites with the same sandbox, therefore we must generate
```

Comment:
Is this `TODO` note still valid?

Reply:
Good point. This is actually not a todo but a generally valid statement. I'll rephrase the comment.

Reply:
Addressed in 9bf50c0.